plugins/system/languagefilter/src/Extension/LanguageFilter.php

```diff
@@ -327,171 +327,115 @@ public function postprocessNonSEFBuildRule(&$router, &$uri)
-                if ($lang_code === $this->default_lang && $this->params->get('remove_default_prefix', 0)) {
-                    // Create a cookie.
-                    $this->setLanguageCookie($lang_code);
-
-                    $found = false;
-                    array_shift($parts);
-                    $path = implode('/', $parts);
-                }
-
-                // We have found our language and the first part of our URL is the language prefix
-                if ($found) {
-                    array_shift($parts);
-
-                    // Empty parts array when "index.php" is the only part left.
-                    if (\count($parts) === 1 && $parts[0] === 'index.php') {
-                        $parts = [];
-                    }
-
-                    $uri->setPath(implode('/', $parts));
-                }
+            if (isset($this->sefs[$sef])) {
+                // We found a matching language to the lang code
+                $uri->setVar('lang', $this->sefs[$sef]->lang_code);
+                array_shift($parts);
+                $uri->setPath(implode('/', $parts));
+            } elseif ($this->params->get('remove_default_prefix', 0)) {
+                // We don't have a prefix for the default language
+                $uri->setVar('lang', $this->default_lang);
             }
-        } else {
-            // We are not in SEF mode
-            $lang_code = $this->getLanguageCookie();
+        } elseif ($uri->hasVar('lang')) {
+            // We are not in SEF mode. Do we have a language set?
+            $lang_code = $uri->getVar('lang');
 
-            if (!$lang_code && $this->params->get('detect_browser', 1)) {
-                $lang_code = LanguageHelper::detectLanguage();
-            }
-
-            if (!isset($this->lang_codes[$lang_code])) {
-                $lang_code = $this->default_lang;
+            if (isset($this->sefs[$lang_code])) {
+                // We found a matching language to the lang code
+                $uri->setVar('lang', $this->sefs[$lang_code]->lang_code);
+            } else {
+                // The language is not installed on our site
+                $uri->delVar('lang');
             }
         }
+    }
 
-        $lang = $uri->getVar('lang', $lang_code);
+    /**
+     * Parse rule to set the applications language state.
+     * This rule is removed after being executed the first time, since
+     * it does redirects and thus disallows parsing more than one URL per page call
+     *
+     * @param   Router  &$router  Router object.
+     * @param   Uri     &$uri     Uri object.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function setLanguageApplicationState(&$router, &$uri)
+    {
+        // We check if the parseRule is still attached to keep this b/c
+        if (!in_array([$this, 'parseRule'], $router->getRules()['parsepreprocess'])) {
```

Comment:
```suggestion
        if (!\in_array([$this, 'parseRule'], $router->getRules()['parsepreprocess'])) {
```
php-cs-fixer suggestion 1 of 2